src/software/ai/hl/stp/play/dynamic_plays/support_tactic_scorers/support_tactic_candidate.tpp

```diff
@@ -0,0 +1,62 @@
+#include "software/ai/hl/stp/play/dynamic_plays/support_tactic_scorers/support_tactic_candidate.h"
+
+template<typename TSupportTactic>
+SupportTacticCandidate<TSupportTactic>::SupportTacticCandidate() 
+    : scores_(), total_score_(0.0), total_score_invalidated_(false)
+{
+}
+
+template<typename TSupportTactic>
+void SupportTacticCandidate<TSupportTactic>::score(SupportTacticScorer &scorer)
+{
+    // We don't care about scores outside the range [-1.0, 1.0]
+    double score = std::clamp(scorer.score(*this), -1.0, 1.0);
+    scores_.push_back(score);
+
+    // Indicate that total score needs to be recomputed
+    total_score_invalidated_ = true;
+}
+
+template<typename TSupportTactic>
+void SupportTacticCandidate<TSupportTactic>::resetTotalScore()
+{
+    scores_.clear();
+    total_score_ = 0.0;
+}
+
+template<typename TSupportTactic>
+double SupportTacticCandidate<TSupportTactic>::getTotalScore()
+{
+    if (total_score_invalidated_) 
+    {
+        computeTotalScore();
+        total_score_invalidated_ = false;
+    }
+    
+    return total_score_;
+}
```

Comment:
why does the support_tactic_candidate have to keep track of the score? Couldn't the play keep track of scorers since it will ultimately decide which tactic to run?

Reply:
When we use `std::max_element` to find the candidate with the highest score, the comparison function will call `getTotalScore` to compare the current best candidate and the next candidate in the iterator. So some candidates may have `getTotalScore` called on it multiple times. I'm caching the score for each candidate within the candidate itself so we don't recalculate it over and over